Allow developers to provide their own entity id

dfederm opened the pull request:
This change allows developers to provide their own entity id as opposed to just always using the content url as the unique identifier.

This allows developers to use multiple urls which may have minor differences (param ordering, presentation-related query params, etc.) which all should be considered to point to the same underlying entity. This also makes it easier for things like bots to create deep links to existing tabs without knowing the exact content url for example.

As a reminder, we had originally decided that we wouldn't need an entity id since we could generate it from the content url, and if we had the content url we wouldn't need an entity id, but this makes assumptions about a url exactly and uniquely identifying an entity, which as described above isn't necessarily a good assumption.

ShreeBalasubramanian commented on src/MicrosoftTeams.ts:
```diff
@@ -875,6 +881,12 @@ namespace microsoftTeams
+         * The developer-defined unique id for the entity this content points to.
+         * If none was provided, this value is generated from the contentUrl.
+         */
+        entityId: string;
```
why not optional here ?

dfederm replied:
We'll always provide it. It may just be a hash of the content url though at this point, but it'll always be given.

dfederm commented on src/MicrosoftTeams.ts:
```diff
@@ -872,7 +870,18 @@ namespace microsoftTeams
          * The Microsoft Teams id for the channel with which the content is associated.
          */
-        channelId?: string;
+        channelId: string;
```
Based on discussion with Yuri, this will always be provided by the client.

dfederm commented:
Current summary of the changes:
- Replaced customSettings with entityId and expose it as part of the Context (as opposed to only the settings frame)
- Replaced deepLinkContext with subEntityId
- Made label required for deep links

ydogandjiev commented on src/MicrosoftTeams.ts:
```diff
@@ -330,12 +331,9 @@ namespace microsoftTeams
             websiteUrl?: string;
 
             /**
```
Wait, are we completely getting rid of customSettings?

dfederm replied:
Yup, it's being replaced with entity id. I know they're not entirely the same thing, but they serve the same basic intent. Entity Id is actually slightly more powerful since it'll be exposed to the content frame and not just the settings frame.

richardtaylorrt replied:
Do we still document the 1kb restriction somewhere?

ydogandjiev commented on test/MicrosoftTeams.spec.ts:
```diff
@@ -444,6 +448,7 @@ describe("MicrosoftTeams", () =>
             contentUrl: "someContentUrl",
             websiteUrl: "someWebsiteUrl",
             customSettings: "someCustomSettings",
```
If customSettings are going away then you should delete this.

dfederm commented on src/MicrosoftTeams.ts:
```diff
@@ -210,7 +210,8 @@ namespace microsoftTeams
-            deepLinkParameters.deepLinkContext,
+            deepLinkParameters.entityId,
+            deepLinkParameters.subEntityId,
             deepLinkParameters.label,
```
This is a breaking change but we haven't released this yet.

dfederm commented on src/MicrosoftTeams.ts:
```diff
@@ -210,7 +210,8 @@ namespace microsoftTeams
         sendMessageRequest(parentWindow, "shareDeepLink", [
-            deepLinkParameters.deepLinkContext,
+            deepLinkParameters.entityId,
+            deepLinkParameters.subEntityId,
```
Hmm, we actually don't need this since it's this tab's entity id.

richardtaylorrt commented on src/MicrosoftTeams.ts:
```diff
@@ -900,25 +908,20 @@ namespace microsoftTeams
          * The label which should be displayed for this deep link when the link is rendered in a client.
          */
-        label?: string;
+        label: string;
```
I wonder if this should be subEntityLabel and subEntityWebUrl?  Or if the interface should be SubEntityDeepLinkParameters?  To avoid confusion elsewhere with the main (pinned) entity's label and webUrl.

richardtaylorrt replied:
...and does the resulting deeplink contain both the labels as it contains both the IDs?  Might need the entity label itself if user decides to pin it as a tab.

dfederm replied:
I don't think we need those. I think we're expecting the sub-entity to be renderable using the parent entity url + subEntityId provided via the SDK. Else we're going to start hitting the url length limit very soon with 2 full urls.

dfederm replied:
The label is for the deeplink, not the entity right?
